### PR TITLE
Proposal:  New bind operation `Field` to specifically construct fields of a `Bundle`

### PR DESCRIPTION
**Pull request by chick**
This has bounced around for a while now. Let's get some feedback and take action

### CODE REVIEW

File: text/000-bundle-field-api.md [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+  val x = Field(UInt(32.W))
+  val y = Field(UInt(16.W))
+  val flag = Field(Bool())
+  val constant = 7.U
+}
+```
+
+The new semantics are, tentatively:
+
+`Fields` are all or nothing, if you define any `Bundle` element as a `Field`, you must define all Bundle elements as Fields. The compiler will error out if you either have vals of type Data that aren't declared as a Field, or if you have Fields that weren't autodetected to be Bundle fields (which is a limitation of reflection-based detection).
```
Comment (chick) [inline text/000-bundle-field-api.md]: Is this analysis possible? If there are hardware elements that are used a generators they will not be wrapped in Field.

Comment (davidbiancolin) [reply text/000-bundle-field-api.md]: Yeah i kinda thought the point of this was, in part, to help disambiguate between these two cases, not to forbid having non-field data members altogether. I'm not sure it is sufficient to just support the gen and constant cases (?). Seems to me like we also need a `NotAField`?

**Comment by azidar** [comment]
Can we merge this PR, as its just an RFC?

**Comment by chick** [comment]
@azidar I've put that question on the Chisel agenda. I feel like there ought to be some formal decision on that process

File: text/000-bundle-field-api.md [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+Suggestions are welcome!
+
+# Background
+[background]: #background
+
+The standard pattern for using Chisel Data objects as generator parameters in `Bundle`s is to pass 
+them as a constructor argument like so:
+```scala
+class MyBundle[T <: Data](gen: T) extends Bundle {
+  val foo = Input(gen)
```
Comment (johnsbrew) [inline text/000-bundle-field-api.md]: you need a chiselTypeOf() or .cloneType here

**Comment by seldridge** [comment]
Is there anything that we currently can't do with just approach (1) and telling users to use `private val` constructor parameters?

I don't know if the current situation is that bad given that there's only one lemma to remember:

- Only public vals of type `Data` or `Some[Data]` will show up as hardware fields in the bundle

I'm not a huge fan of the all-or-nothing approach where the stateful use of `Field` results in different semantics. I think it may make more sense to push people onto full use of `Field` (which may require a new class like `Struct`?). Also, does the all-or-nothing state extend across `Bundle`s inside of `Bundle`s?

It occurred to me that `Field` and `IO` are very similar---they promote public `val`s of type `Data` to be something that you can connect to. Are these really the same thing with different contexts (the former being inside a `Bundle` and the latter being inside a `Module`)?

File: text/000-bundle-field-api.md (hunk 2 above)
Comment (jackkoenig) [reply text/000-bundle-field-api.md]: You only need a `chiselTypeOf` if you're intending to pass bound hardware as `gen`, but the standard style is to use unbound types so this is correct.

**Comment by azidar** [comment]
Let's vote!!

Vote yay: Thumbs up
Vote nay: Thumbs down

Clarification - this is voting on whether to work on an implementation (advancing to the next stage). We will then discuss implementation.

**Comment by schoeberl** [comment]
Does the change to "### 2. Bundle Elements should be vals in the *body* of a Bundle" resolve the issue and we do not need to add Field?

**Comment by jackkoenig** [comment]
> Does the change to "### 2. Bundle Elements should be vals in the body of a Bundle" resolve the issue and we do not need to add Field?

It doesn't resolve field aliasing:
```scala
class MyBundle(gen: UInt) extends Bundle {
  val a = gen
  val b = gen
}
```
This can be handled using by-name arguments like so:
```scala
class MyBundle(gen: => UInt) extends Bundle {
  val a = gen
  val b = gen
}
```
But this also can be problematic if used improperly:
```scala
// This is proper use and will work
new MyBundle(UInt(8.W))
// This has the same issue as not using by-name arguments
val x = UInt(8.W)
new MyBundle(x)
```

**Pull request merged**